File: 238.product-of-array-except-self.py
#
# @lc app=leetcode id=238 lang=python3
#
# [238] Product of Array Except Self
#

# @lc code=start
from itertools import accumulate
import operator
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # pre[i]: prefix prod <= i
        # suf[i]: suffix prod >= i
        pre = list(accumulate(nums,operator.mul))
        suf = list(accumulate(nums[::-1], operator.mul))[::-1]

        res = []
        for i in range(len(nums)):
            if i==0:
                res.append(suf[i+1])
            elif i==len(nums)-1:
                res.append(pre[i-1])
            else:
                res.append(suf[i+1]*pre[i-1])
        return res

# @lc code=end

File: 238.product-of-array-except-self.go
/*
 * @lc app=leetcode id=238 lang=golang
 *
 * [238] Product of Array Except Self
 */

// @lc code=start
func productExceptSelf(nums []int) []int {
	n := len(nums)

	prefixProduct := make([]int, n)
	acc := 1
	for i := 0; i < n; i++ {
		prefixProduct[i] = acc
		acc *= nums[i]
	}

	suffixProduct := make([]int, n)
	acc = 1
	for i := n - 1; i >= 0; i-- {
		suffixProduct[i] = acc
		acc *= nums[i]
	}

	ans := make([]int, n)
	for i := 0; i < n; i++ {
		ans[i] = suffixProduct[i] * prefixProduct[i]
	}
	return ans
}

// @lc code=end

